app/controllers/email_confirmations_controller.rb

```diff
@@ -75,4 +112,32 @@ def email_params
   def token_params
     params.permit(:token).require(:token)
   end
+
+  def setup_mfa_authentication
+    return if @user.mfa_disabled?
+    @form_mfa_url = mfa_update_email_confirmations_url(token: @user.confirmation_token)
+  end
+
+  def setup_webauthn_authentication
+    return if @user.webauthn_credentials.none?
+
+    @form_webauthn_url = webauthn_update_email_confirmations_url(token: @user.confirmation_token)
+
+    @webauthn_options = @user.webauthn_options_for_get
+
+    session[:webauthn_authentication] = {
+      "challenge" => @webauthn_options.challenge,
+      "user" => @user.id
+    }
+  end
+
+  def webauthn_verification_failure(message = nil)
```

Comment:
Dumb question: Why can it be nil?

Reply:
Also: in my reading of the code, if I pass a nil, then it defaults to nil. I'm not sure what this line is intended to do.

Reply:
Not sure if this is what you mean, but this is the default value for if no param is passed in. So if I call `webauthn_verification_failure` without providing a message, then message will be nil. This isn't explicitly needed here, as I am always providing a message, but it helps to keep the function 'generic'.